pkg/sync/sync.go

```diff
@@ -1,93 +1,95 @@
-	s3Path =  strings.Replace(s3Path, "s3://", "", -1)
-	watcher, err := fsnotify.NewWatcher()
-	if err != nil {
-		return nil, err
-	}
+func NewSyncer(fchan chan notify.EventInfo, folderToWatch, s3Region, s3Path string, stop chan struct{}, deleteFile bool) (Syncer, error) {
+	s3Path = strings.Replace(s3Path, "s3://", "", -1)
+	// Make the channel buffered to ensure no event is dropped. Notify will drop
+	// an event if the receiver is not able to keep up the sending pace.
 	s3session, err := session.NewSession(&aws.Config{Region: aws.String(s3Region)})
 	if err != nil {
 		return nil, err
 	}
 	uploader := s3manager.NewUploader(s3session)
 	return &syncer{
-		folderToWatch: folderToWatch,
-		s3Bucket:      strings.Split(s3Path, "/")[0],
-		s3Folder:      strings.Join(strings.Split(s3Path, "/")[1:], "/"),
-		stop:          stop,
-		watcher:       watcher,
-		s3session:     s3session,
-		uploader:      uploader,
+		fchan:      fchan,
+		s3Bucket:   strings.Split(s3Path, "/")[0],
+		s3Folder:   strings.Join(strings.Split(s3Path, "/")[1:], "/"),
+		stop:       stop,
+		s3session:  s3session,
+		uploader:   uploader,
+		deleteFile: deleteFile,
 	}, nil
 }
 
+func IsDirectory(path string) bool {
+	fileInfo, err := os.Stat(path)
+	if err != nil {
+		return false
+	}
+	return fileInfo.IsDir()
+}
+
 func (s *syncer) Sync() error {
-	go func() {
-		for {
-			select {
-			// watch for events
-			case event := <-s.watcher.Events:
-				log.Debug(event)
-				// watch for errors
-				switch {
-				case event.Op&fsnotify.Write == fsnotify.Write:
-					log.Infof("uploading modified file '%s' to S3", event.Name)
-					go s.uploadFile(event.Name)
-				case event.Op&fsnotify.Create == fsnotify.Create:
-					log.Infof("uploading new file '%s' to S3", event.Name)
-					go s.uploadFile(event.Name)
-				default:
-					log.Debug("no changes")
-				}
-			case err := <-s.watcher.Errors:
-				log.Error(err)
-			case <-s.stop:
-				s.watcher.Close()
+	for {
+		select {
+		case event := <-s.fchan:
+			log.Debug(event.Path())
+			switch {
+			// watch for errors
+			case IsDirectory(event.Path()):
+				log.Debugf("new directory '%s'", event.Path())
+			case event.Event() == notify.Write:
+				log.Infof("uploading modified file '%s' to S3", event.Path())
+				go s.uploadFile(event.Path(), s.deleteFile)
+			case event.Event() == notify.Create:
+				log.Infof("uploading new file '%s' to S3", event.Path())
+				go s.uploadFile(event.Path(), s.deleteFile)
+			default:
+				log.Debug("no changes")
 			}
 		}
-	}()
-
-	// Add the folder to watch to the Watcher
-	if err := s.watcher.Add(s.folderToWatch); err != nil {
-		return err
 	}
-
-	return nil
 }
 
-func (s *syncer) uploadFile(fileDir string) {
+func (s *syncer) uploadFile(fileDir string, deleteFile bool) {
 	// Upload
 	err := s.addFileToS3(s.s3session, fileDir)
 	if err != nil {
 		log.Error(err)
+	} else {
+		if deleteFile {
+			var errDelete = os.Remove(fileDir)
+			if os.IsNotExist(errDelete) {
```

Comment:
I would say that cases opposite to this are even more important to log

Reply:
I've removed deleting part at the end. I showed to be more tricky than expected.